// rebass/dist/Panel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Panel = undefined;

var _components = require('@rebass/components');

var _components2 = _interopRequireDefault(_components);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Panel = exports.Panel = (0, _components2.default)({
  bg: 'white',
  borderRadius: 2,
  border: 1,
  borderColor: 'gray'
}, {
  overflow: 'hidden'
}, 'space', 'color', 'borders', 'borderColor', 'borderRadius');

Panel.displayName = 'Panel';

Panel.Header = (0, _components2.default)({
  is: 'header',
  fontSize: 2,
  p: 2,
  bg: 'white',
  borderBottom: 1,
  borderColor: 'gray'
}, 'space', 'color', 'fontSize', 'borders', 'borderColor');

Panel.Header.displayName = 'Panel.Header';

Panel.Footer = (0, _components2.default)({
  is: 'footer',
  fontSize: 1,
  p: 2,
  bg: 'white',
  borderTop: 1,
  borderColor: 'gray'
}, 'space', 'colors', 'fontSize', 'borders', 'borderColor');

Panel.Footer.displayName = 'Panel.Footer';

exports.default = Panel;
